src/rust/session/src/buffer.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright 2021 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+//! A very simple buffer type that can be replaced in the future.
+
+use core::borrow::{Borrow, BorrowMut};
+
+/// A growable byte buffer
+pub struct Buffer {
+    buffer: Vec<u8>,
+    position: usize,
+    capacity: usize,
+    target_capacity: usize,
+}
+
+impl Buffer {
+    /// Create a new `Buffer` that can hold up to `capacity` bytes without
+    /// re-allocating.
+    pub fn with_capacity(capacity: usize) -> Self {
+        let mut buffer = Vec::with_capacity(capacity);
+        buffer.resize(capacity, 0);
+
+        Self {
+            buffer,
+            capacity: 0,
```

Comment:
why not preallocate when session is created?

Reply:
This is a somewhat misleading name, I did this to match `BufReader` https://doc.rust-lang.org/src/std/io/buffered/bufreader.rs.html#48-53

Arguably, position is really a read offset, and 'capacity' is serving as a write offset. The buffer's length is the distance between the two and the actual capacity is the capacity of the inner buffer (`Vec<u8>`).

I've pushed some changes which I'm not sure if you saw at the time of this review. Take another look and I can update the naming.

Reply:
Renamed these fields to make it a little less confusing.